src/ff1.rs

```diff
@@ -0,0 +1,459 @@
+            let s = generate_s(&self.ciph, &r[..], d);
+
+            // 6iv. Let y = NUM(S).
+            let y = BigUint::from_bytes_be(&s);
+
+            // 6v. If i is even, let m = u; else, let m = v.
+            let m = if i % 2 == 0 { u } else { v };
+
+            // 6vi. Let c = (NUM(A, radix) + y) mod radix^m.
+            let c = (num_radix(&x_a, &self.radix_bi) + y) % pow(&self.radix_bi, m);
+
+            // 6vii. Let C = STR(c, radix).
+            let x_c = str_radix(c, self.radix, m);
+
+            // 6viii. Let A = B.
+            x_a = x_b;
+
+            // 6ix. Let B = C.
+            x_b = x_c;
+        }
+
+        // 7. Return A || B.
+        [x_a, x_b].concat()
+    }
+
+    pub fn decrypt(&self, tweak: &[u8], x: &[Radix]) -> Vec<Radix> {
+        let n = x.len();
+        let t = tweak.len();
+
+        // 1. Let u = floor(n / 2); v = n - u
+        let u = n / 2;
+        let v = n - u;
+
+        // 2. Let A = X[1..u]; B = X[u + 1..n].
+        let mut x_a = Vec::from(&x[0..u]);
+        let mut x_b = Vec::from(&x[u..n]);
+
+        // 3. Let b = ceil(ceil(v * log2(radix)) / 8).
+        let b = (v as f64 * (self.radix as f64).log2() / 8f64).ceil() as usize;
+
+        // 4. Let d = 4 * ceil(b / 4) + 4.
+        let d = (4f64 * (b as f64 / 4f64).ceil() + 4f64) as usize;
+
+        // 5. Let P = [1, 2, 1] || [radix] || [10] || [u mod 256] || [n] || [t].
+        let mut p = vec![1, 2, 1];
+        p.write_u24::<BigEndian>(self.radix as u32).unwrap();
+        p.write_u8(10).unwrap();
+        p.write_u8(u as u8).unwrap();
+        p.write_u32::<BigEndian>(n as u32).unwrap();
+        p.write_u32::<BigEndian>(t as u32).unwrap();
+
+        //  6i. Let Q = T || [0]^((-t-b-1) mod 16) || [i] || [NUM(A, radix)].
+        let q_base = {
+            let val = ((((-(t as i32) - (b as i32) - 1) % 16) + 16) % 16) as usize;
+            let mut q = Vec::from(tweak);
+            q.resize(t + val, 0);
+            q
+        };
+        for i in 0..10 {
+            let i = 9 - i;
+            let mut q = q_base.clone();
+            q.write_u8(i).unwrap();
+            let q_bytes = num_radix(&x_a, &self.radix_bi).to_bytes_be();
+            for _ in 0..(b - q_bytes.len()) {
+                q.write_u8(0).unwrap();
+            }
+            q.extend(q_bytes);
+
+            // 6ii. Let R = PRF(P || Q).
+            let r = self.prf(&[&p[..], &q[..]].concat());
+
+            // 6iii. Let S be the first d bytes of R.
+            let s = generate_s(&self.ciph, &r[..], d);
+
+            // 6iv. Let y = NUM(S).
+            let y = BigInt::from(BigUint::from_bytes_be(&s));
+
+            // 6v. If i is even, let m = u; else, let m = v.
+            let m = if i % 2 == 0 { u } else { v };
+
+            // 6vi. Let c = (NUM(B, radix) - y) mod radix^m.
+            let modulus = BigInt::from(pow(&self.radix_bi, m));
+            let mut c = (BigInt::from(num_radix(&x_b, &self.radix_bi)) - y) % &modulus;
```

Comment:
BTW, a criticism of the algorithm here (not the implementation, which is correct): In the case of radix 2 this is fine (although overcomplicated) because the modulus (*radix*<sup>*m*</sup> in the spec) will be a multiple of 2<sup>ceiling(*n*/2)</sup>. But for non-power-of-2 radices, the 4 extra bytes in *d*, mentioned in footnote 6, might not be enough to eliminate cryptanalytically useful bias in the reduction.

Also, taking the algorithm literally results in computing an unnecessary extra block-encryption per PRF evaluation when *radix* is a power of 2 and ceiling(*n*/2).lg(*radix*) mod 128 > 96.